app/uk/gov/hmrc/tai/util/IORetryExtension.scala

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright 2023 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.tai.util
+
+import cats.effect.IO
+
+import scala.concurrent.duration.FiniteDuration
+
+class LockedException(message: String) extends Exception(message)
+
+object IORetryExtension {
+  implicit class Retryable[A](io: IO[A]) {
+    def simpleRetry(noOfRetries: Int, sleep: FiniteDuration): IO[A] = {
+      def retryLoop(times: Int): IO[A] = {
+        io.map(identity).handleErrorWith {
+          case _: LockedException if times != 0 =>
+            IO.sleep(sleep) >> retryLoop(times - 1)
```

Comment:
I don't know how deep this recursion will end up but should it not be made tail recursive and annotated with @tailrec?

Reply:
max depth is 20. I have tried to write a tail recursive version without success.

Reply:
OK, I think the problem is the retry happens inside the IO object but I'm not sure how to get round that, not having used IO before. I'll approve for now, but it might be worth raising a Jira ticket to look into this at some point as we don't want to be running out of stack space.